library/junos_pmtud

```diff
@@ -0,0 +1,263 @@
+  max_range:
+    description:
+        - Max range of path MTU discovery. Must be 2^n.
+    required: false
+    default: 512
+'''
+
+EXAMPLES = '''
+# Simple example
+  tasks:
+    - name: "Check MTU on backup circuit"
+      junos_pmtud:
+        host={{ junos_host }}
+        port={{ netconf_port }}
+        user={{ ansible_ssh_user }}
+        passwd={{ ansible_ssh_pass }}
+        dest_ip=8.8.8.8
+
+# Using more parameters
+  tasks:
+    - name: "Check MTU on backup circuit"
+      junos_pmtud:
+        host={{ junos_host }}
+        port={{ netconf_port }}
+        user={{ ansible_ssh_user }}
+        passwd={{ ansible_ssh_pass }}
+        dest_ip=8.8.8.8
+        routing_instance=internet
+        max_range=128
+'''
+
+from distutils.version import LooseVersion
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            host=dict(required=True),
+            user=dict(required=False, default=os.getenv('USER')),
+            passwd=dict(required=False, default=None, no_log=True),
+            port=dict(required=False, default=830),
+            ssh_private_key_file=dict(required=False, default=None),
+            mode=dict(required=False, default=None),
+            timeout=dict(required=False, default=0),
+            dest_ip=dict(required=True, default=None),
+            source_ip=dict(required=False, default=None),
+            interface=dict(required=False, default=None),
+            routing_instance=dict(required=False, default=None),
+            max_size=dict(required=False, default=1472),
+            max_range=dict(required=False, default=512),
+            ),
+        supports_check_mode=False
+    )
+
+    m_args = module.params
+
+    try:
+        from jnpr.junos import Device
+        from jnpr.junos.version import VERSION
+        if not LooseVersion(VERSION) >= LooseVersion('1.2.2'):
+            module.fail_json(msg='junos-eznc >= 1.2.2 is required for this module')
+    except ImportError as ex:
+        module.fail_json(msg='ImportError: %s' % ex.message)
+
+    if m_args['mode'] is not None and LooseVersion(VERSION) < LooseVersion('2.0.0'):
+        module.fail_json(msg='junos-eznc >= 2.0.0 is required for console connection')
+
+    # Open connection to device
+    dev = Device(
+            m_args['host'],
+            user=m_args['user'],
+            passwd=m_args['passwd'],
+            port=m_args['port'],
+            ssh_private_key_file=m_args['ssh_private_key_file'],
+            mode=m_args['mode'],
+            gather_facts=False)
+
+    try:
+        dev.open()
+    except Exception as err:
+        msg = 'Unable to connect to {0}: {1}'.format(m_args['host'], str(err))
+        module.fail_json(msg=msg)
+        return
+    results = dict(
+        changed=False,
+        mtu=0,
+        dest_ip=m_args['dest_ip']
+    )
+
+    # Prepare parameters
+    ping_params = dict(
+            host=m_args['dest_ip'],
+            count='3',
+            do_not_fragment=True,
```

Comment:
As it currently stands, a user could enter an IPv6 address for `dest-ip` (or a hostname which resolves to an IPv6 address) and the module will return a value that is 20 bytes less than the actual IPv6 MTU because it assumes IPv4 sized header.

For now, we can force IPv4 only and avoid an incorrect answer by adding `inet=True` to `ping_params`

In the long run, additional arguments and code could be added for other non-IPv4 address families.

Reply:
Admittedly, I had only IPv4 in mind when writing this. At the moment I don't have environment with IPv6 running on JUNOS so I agree to limit the module to IPv4 now.